### ISSUE TITLE
A flag to enable the previous RNG behavior for `tf.keras.initializers`

### ISSUE THREAD

**Issue by ydshieh** [tensorflow/tensorflow#57763]
### Issue Type

Feature Request

### Source

binary

### Tensorflow Version

TF 2.10

### Custom Code

No

### Current Behaviour?

According to [release note](https://github.com/tensorflow/tensorflow/releases/tag/v2.10.0), RNG behavior change for `tf.keras.initializers` since `TF 2.10`.

In particular, `tf.random.set_seed(seed)` is not enough to get the same model weights - we need to set explicit seeds in the initializers.

In general, this is NOT an issue. However, for **testing purpose**, this makes things much more difficult. For example, let's say we have `MyCustomModel`, implemented with layers without specifying any initializer (therefore, no explicit seed with initializer).

However, if we want to have a CI test that verifies the model produces the same result, it is impossible with TF 2.10 due to the new change. With TF < 2.10, we can set `tf.random.set_seed(seed=0)`, and the model weights will be the same, so we can check if the output matches the expected values.

It would be great if there is a way to have previous behavior for **testing purpose**.

**Comment by ydshieh** [comment]
Not sure who to tag, but @qlzh727 had a commit 23d4b96c that is related to this change.

**Comment by edwardyehuang** [comment]
Agree. We need a flag or a function to enable/disable the stateless/stateful rng_type

**Comment by qlzh727** [comment]
Thanks for reporting this issue. As a short cut for now (to have a determined initializer for your test), you can use `tf.keras.utils.set_random_seed()` to set the seed for python/numpy/tf/keras.

**Comment by edwardyehuang** [comment]
> Thanks for reporting this issue. As a short cut for now (to have a determined initializer for your test), you can use `tf.keras.utils.set_random_seed()` to set the seed for python/numpy/tf/keras.

Hi. I think you misunderstand it. Even set random seed using `tf.keras.utils.set_random_seed()`, the generated weights is different between 2.10 and 2.9/2.8 (Stateless vs Stateful rng)

**Comment by qlzh727** [comment]
I see. Is it because there are some golden number that is baked in the test case? (I think TF doesn't have a contract for the exact RNG number created  between version. Adding @wangpengmit for this.)

**Comment by ydshieh** [comment]
Yes, we fix some inputs, and use a seed to create models -> and test the model outputs will match the pre-defined expected values.

> TF doesn't have a contract for the exact RNG number created between version.

I understand - same situation with PyTorch. So far (up to TF 2.9) our tests are fine as we still get the same model weights. But with TF 2.10, some tests are failing. We probably need to re-consider our testing approach, but still wonder if TensorFlow could provide some mechanism to make things much easier 🙏

**Comment by jxy** [comment]
The behavior of the stateless RNG initializers encumbers reproducibility.  As `tf.random.set_seed` no longer affects the initializers, and the only way to have a reproducible initializer is explicit passing a seed at the call site during initialization, there is no easy way to change training runs from using different initial weights to reproducible weights.

If reverting the initializer behavior is not the way forward, we would like to see a new api that change the default seed of the initializers globally, without touching existing model code.

**Comment by jxy** [comment]
> The global seed setting for keras initializer is tf.keras.utils.set_random_seed(), which should make all the initializer to have a determined seed.

Indeed.  Thanks.

The doc string https://github.com/keras-team/keras/blob/b80dd12da9c0bc3f569eca3455e77762cf2ee8ef/keras/utils/tf_utils.py#L44-L53 needs to be updated to reflect that it's also setting https://github.com/keras-team/keras/blob/b80dd12da9c0bc3f569eca3455e77762cf2ee8ef/keras/utils/tf_utils.py#L66

**Comment by ydshieh** [comment]
Thank a lot! `tf.keras.utils.set_random_seed()` makes TF fans' life much easier :-)

**Comment by qlzh727** [comment]
Ack. Closing this issue now.

**Comment by edwardyehuang** [comment]
@qlzh727

I think providing a flag for the switch between stateless and stateful is still necessary. I found the new RNG behavior (same seed, same hardware with deterministic enabled) produced a different result. Check this link (https://www.tensorflow.org/guide/random_numbers). It states, "The RNG algorithm used by stateless RNGs is device-dependent, meaning the same op running on a different device may produce different outputs."
